calyx/src/analysis/live_range_analysis.rs

```diff
@@ -6,13 +6,70 @@ use itertools::Itertools;
 use std::{
     collections::{HashMap, HashSet},
     fmt::Debug,
+    rc::Rc,
 };
 
 type TypeNameSet = HashSet<(ir::CellType, ir::Id)>;
 type CellsByType = HashMap<ir::CellType, HashSet<ir::Id>>;
 // maps cell type to maps that map cell name to control statement
 type LiveMapByType = HashMap<ir::CellType, HashMap<ir::Id, HashSet<u64>>>;
 
+/// Returns [ir::Cell] which are read from in the assignments.
+/// **Ignores** reads from group holes, and reads from done signals, when it
+/// is safe to do so.
+/// To ignore a read from a done signal:
+/// the `@go` signal for the same cell *must* be written to in the group
+pub fn meaningful_read_set<'a>(
+    assigns: impl Iterator<Item = &'a ir::Assignment> + Clone + 'a,
+) -> impl Iterator<Item = RRC<ir::Cell>> + 'a {
+    meaningful_port_read_set(assigns)
+        .map(|port| Rc::clone(&port.borrow().cell_parent()))
+        .unique_by(|cell| cell.clone_name())
+}
+
+/// Returns the "meaningful" [ir::Port] which are read from in the assignments.
+/// "Meaningful" means we just exclude the following `@done` reads:
+/// the `@go` signal for the same cell *must* be written to in the group
+pub fn meaningful_port_read_set<'a>(
```

Comment:
Can we move this out from the analysis (which is supposed to only provide generic operations)? One done, merge it!

Reply:
Hmmmm, I actually use `meaningful_port_read_set()` in live range analysis, since only the idea of "meaningful port reads" (as opposed to all reads) is what we want to use when constructing a live range analysis. Where else could I put it?

Reply:
Ah, my bad. I thought this was in read_write_set.rs. You're all good!